File: search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        for (int i = 0; i < 60; ++i) {
            int mid = (left + right) / 2;
            if (nums[mid] == target || nums[left] == target || nums[right] == target) {
                if (nums[mid] == target) return mid;
                if (nums[left] == target) return left;
                if (nums[right] == target) return right;
            }
            if (nums[mid] > nums[left]) {
                if (target > nums[left] && target < nums[mid]) {
                    right = mid;
                } else {
                    left = mid;
                }
            } else {
                if (target > nums[left] || target < nums[mid]) {
                    right = mid;
                } else {
                    left = mid;
                }
            }
        }
        return -1;
    }
};

File: search-in-rotated-sorted-array.java
```
public class Solution {

    public int search(int[] nums, int target) {
        int left = 0, right = nums.length-1;
        for (int i = 0; i < 65; ++i) {
            int mid = (left + right) / 2;
            if (nums[left] == target || nums[right] == target || nums[mid] == target) {
                return nums[left] == target ? left : (nums[right] == target ? right : mid);
            }
            if (nums[left] < nums[mid]) {
                if (target > nums[left] && target < nums[mid]) {
                    right = mid;
                } else {
                    left = mid;
                }
            } else {
                if (target > nums[left] || target < nums[mid]) {
                    right = mid;
                } else {
                    left = mid;
                }
            }
        }
        return -1;

    }
}
```